File: android/src/main/java/ch/nadlo/oss/capacitor/pdf_viewer/PDFViewerPlugin.java
package ch.nadlo.oss.capacitor.pdf_viewer;

import android.app.Activity;

import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "PDFViewer")
public class PDFViewerPlugin extends Plugin {
    private PdfViewerFragment pdfFragment;

    @PluginMethod
    public void open(PluginCall call) {
        String url = call.getString("url");
        Integer top = call.getInt("top", 0);

        if (url != null && !url.isEmpty()) {
            // Obtener la actividad actual
            Activity activity = getActivity();

            if (!(activity instanceof FragmentActivity)) {
                call.reject("Activity is not a FragmentActivity");
                return;
            }

            FragmentActivity fragmentActivity = (FragmentActivity) activity;
            FragmentManager fragmentManager = fragmentActivity.getSupportFragmentManager();

            // Crear una nueva instancia del fragmento y mostrarlo
            fragmentActivity.runOnUiThread(() -> {
                pdfFragment = PdfViewerFragment.newInstance(url, top);
                fragmentManager.beginTransaction()
                        .replace(android.R.id.content, pdfFragment)  // Usa el contenedor predeterminado
                        .addToBackStack(null)  // Permite volver atrás
                        .commit();

                call.resolve();
            });
        }
    }

    @PluginMethod
    public void close(PluginCall call) {
        // Obtener la actividad actual
        Activity activity = getActivity();
        if (!(activity instanceof FragmentActivity)) {
            call.reject("Activity is not a FragmentActivity");
            return;
        }

        FragmentActivity fragmentActivity = (FragmentActivity) activity;
        FragmentManager fragmentManager = fragmentActivity.getSupportFragmentManager();

        // Ejecutar en el hilo de la UI para cerrar el fragmento
        fragmentActivity.runOnUiThread(() -> {
            if (pdfFragment != null) {
                fragmentManager.beginTransaction()
                        .remove(pdfFragment)
                        .commit();
                pdfFragment = null; // Limpiar la referencia del fragmento
                call.resolve();
            } else {
                call.reject("No PDF viewer is currently open");
            }
        });
    }
}

File: android/src/main/java/jccham/capacitor/pdf_viewer/PDFViewerPlugin.kt
package jccham.capacitor.pdf_viewer

import androidx.fragment.app.FragmentActivity
import com.getcapacitor.Plugin
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.PluginCall
import androidx.fragment.app.FragmentManager
import com.getcapacitor.PluginMethod

@CapacitorPlugin(name = "PdfViewerPlugin")
class PdfViewerPlugin : Plugin() {
    private var pdfFragment: PdfViewerFragment? = null

    @PluginMethod
    fun loadPdf(call: PluginCall) {
        val pdfUrl = call.getString("url") ?: return call.reject("URL is required")
        val top = call.getInt("top", 0) ?: return call.reject("URL is required")

        // Obtener la actividad actual
        val activity = activity
        if (activity !is FragmentActivity) {
            return call.reject("Activity is not a FragmentActivity")
        }

        val fragmentActivity = activity as FragmentActivity
        val fragmentManager: FragmentManager = fragmentActivity.supportFragmentManager

        // Crear una nueva instancia del fragmento y mostrarlo
        fragmentActivity.runOnUiThread {
            pdfFragment = PdfViewerFragment.newInstance(pdfUrl, top)
            fragmentManager.beginTransaction()
                .replace(android.R.id.content, pdfFragment!!)
                .addToBackStack(null)
                .commit()

            call.resolve()
        }
    }

    @PluginMethod
    fun close(call: PluginCall) {
        val activity = activity
        if (activity !is FragmentActivity) {
            return call.reject("Activity is not a FragmentActivity")
        }

        val fragmentActivity = activity as FragmentActivity
        val fragmentManager: FragmentManager = fragmentActivity.supportFragmentManager

        // Ejecutar en el hilo de la UI para cerrar el fragmento
        fragmentActivity.runOnUiThread {
            if (pdfFragment != null) {
                fragmentManager.beginTransaction()
                    .remove(pdfFragment!!)
                    .commit()
                pdfFragment = null
                call.resolve()
            } else {
                call.reject("No PDF viewer is currently open")
            }
        }
    }
}
